### PR TITLE
feat: add modules to show reusable and parent theme features

### LINKED ISSUE #9655: Add new modules to application theme demo to use a reusable parent theme and application specific theme
- To add a new module to demonstrate how to create a reusable theme that can be packaged as jar
- To add a new module to demonstrate how to reuse a theme as parent (either packaged as jar or existing in another `themes/some-parent` directory)

### PR DESCRIPTION
**Pull request by taefi**
Fixes: https://github.com/vaadin/flow/issues/9655

Following modules added:
- reusable-theme 
- reusing-theme-jar-demo
- extend-existing-app-theme-demo

### CODE REVIEW

**Comment by mshabarov** [comment]
General comment: now, when the demo gets growing gradually, it's useful to add a brief description to `README` file about what is the use case of each module (subproject), +add a link to a corresponding sections in documentation.

File: reusable-theme/src/main/resources/META-INF/resources/themes/custom-theme/view-styles.css [hunk 1]
```diff
@@ -0,0 +1,51 @@
+/*
+These view specific styles are global scoped (like global.css),
```
Comment (mshabarov) [inline reusable-theme/src/main/resources/META-INF/resources/themes/custom-theme/view-styles.css]: `global.css` -> `styles.css`

**Comment by mshabarov** [comment]
Another comment is that the `reusable-theme` contains almost the same styles/files as a theme in `flow-in-app-theme-demo`. Would it better to make it more 'distinct' and to limit `reusable-theme` a bit so as to not bloat it and have only, say, custom background and border colours, border style, and use another base colour (for example, blue, not red/pink).

The modules whose use the reusable theme contains a relatively heavy logic. Is it really necessary? IMO it good to have it in the main module, where you show all the feature of application theming. But here, in these two modules, the goal is to show how to reuse the theme and how to extend it.
So, basically, I have a feeling that a single page with input field and just one button showing a greeting notification would be sufficient for this show case. Otherwise, It's easy to get lost in these views and backends and distinguish theme-reusing relevant code.

WDYT, does it make sense?

File: reusable-theme/pom.xml [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+  <groupId>com.vaadin.demo.apptheme.reusable</groupId>
+  <artifactId>reusable-theme</artifactId>
+  <version>1.0-SNAPSHOT</version>
+
+  <name>reusable-theme</name>
+  <url>https://www.vaadin.com</url>
+
+  <properties>
+    <java.version>11</java.version>
+    <maven.compiler.target>1.8</maven.compiler.target>
```
Comment (mshabarov) [inline reusable-theme/pom.xml:15]: Does it also have to be `11`?

File: reusing-theme-jar-demo/pom.xml [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+        <groupId>org.springframework.boot</groupId>
+        <artifactId>spring-boot-starter-parent</artifactId>
+        <version>2.4.0</version>
+    </parent>
+
+    <pluginRepositories>
+        <pluginRepository>
+            <id>vaadin-prereleases</id>
+            <url>https://maven.vaadin.com/vaadin-prereleases</url>
+            <!--<url>http://tools.vaadin.com/nexus/content/repositories/vaadin-prereleases</url>-->
```
Comment (mshabarov) [inline reusing-theme-jar-demo/pom.xml]: Delete commented code

**Review by mshabarov — CHANGES_REQUESTED** [review]

File: reusable-theme/pom.xml (hunk 2 above)
Comment (taefi) [reply reusable-theme/pom.xml:15]: Done. 
Downgraded to 8.

File: reusable-theme/src/main/resources/META-INF/resources/themes/custom-theme/view-styles.css (hunk 1 above)
Comment (taefi) [reply reusable-theme/src/main/resources/META-INF/resources/themes/custom-theme/view-styles.css]: Done.

File: reusing-theme-jar-demo/pom.xml (hunk 3 above)
Comment (taefi) [reply reusing-theme-jar-demo/pom.xml]: Done.

**Comment by taefi** [comment]
> Another comment is that the `reusable-theme` contains almost the same styles/files as a theme in `flow-in-app-theme-demo`. Would it better to make it more 'distinct' and to limit `reusable-theme` a bit so as to not bloat it and have only, say, custom background and border colours, border style, and use another base colour (for example, blue, not red/pink).

To be honest, I was thinking this would create integrity for this whole demo project. In other words, it is showing that having the same styles in a theme jar, would result in the exact same output as having those resources and styles locally in the project. 

Maybe if I add those README.md files in those projects, it could make it clear. But, anyway, if you think it is better to show a minimal version of the theme packaged as jar, I can apply that.

**Comment by taefi** [comment]
> The modules whose use the reusable theme contains a relatively heavy logic. Is it really necessary? IMO it good to have it in the main module, where you show all the feature of application theming. But here, in these two modules, the goal is to show how to reuse the theme and how to extend it.
> So, basically, I have a feeling that a single page with input field and just one button showing a greeting notification would be sufficient for this show case. Otherwise, It's easy to get lost in these views and backends and distinguish theme-reusing relevant code.

Personally, I think your proposition fits better than having all those codes and logics there, but using the Vaadin-CRM source code for this whole demo thing was Rolf's idea, maybe for being fast, just copy and paste the code and apply the theming, or maybe to show it is working on an already existing project with all those view-specific styles, logics and security almost out-of-the-box. 

For example, I had to add some ignore logic in Spring Security Configs to allow the theming resources to be available, even before authentication. I can't remember if this was the purpose or not. I'm open to doing the changes, but definitely, it needs some time to rework all the modules using Vaadin-CRM code. (currently 3 modules)

**Comment by mshabarov** [comment]
> To be honest, I was thinking this would create integrity for this whole demo project. In other words, it is showing that having the same styles in a theme jar, would result in the exact same output as having those resources and styles locally in the project.
Maybe if I add those README.md files in those projects, it could make it clear. But, anyway, if you think it is better to show a minimal version of the theme packaged as jar, I can apply that.

Ok, thanks for explanation, I got the idea. Let's keep it as it is then. README and docs then would help to focus on the changes needed for reusing themes.

File: extend-existing-app-theme-demo/frontend/themes/specific-theme/components/vaadin-text-field.css [hunk 4]
```diff
@@ -0,0 +1,3 @@
+[part="input-field"] {
+    border: 1px solid green;
+}
```
Comment (mshabarov) [inline extend-existing-app-theme-demo/frontend/themes/specific-theme/components/vaadin-text-field.css:3]: Add a new line

File: extend-existing-app-theme-demo/frontend/themes/specific-theme/styles.css [hunk 5]
```diff
@@ -0,0 +1,10 @@
+html, :host {
+    --lumo-primary-color: green;
+    --lumo-primary-text-color: green;
+    --lumo-font-family: 'Jetbrains Mono';
+}
+
+@font-face {
+    font-family: "Jetbrains Mono";
+    src: url("./some-fonts/jet-brains-mono/JetBrainsMono-Regular.ttf") format("TrueType");
+}
```
Comment (mshabarov) [inline extend-existing-app-theme-demo/frontend/themes/specific-theme/styles.css:10]: Add a new line

File: extend-existing-app-theme-demo/frontend/themes/specific-theme/theme.json [hunk 6]
```diff
@@ -0,0 +1,3 @@
+{
+    "parent": "custom-theme"
+}
```
Comment (mshabarov) [inline extend-existing-app-theme-demo/frontend/themes/specific-theme/theme.json:3]: Add a new line

File: extend-existing-app-theme-demo/pom.xml [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+    <modelVersion>4.0.0</modelVersion>
+
+    <name>Reusing Theme jar Demo</name>
+    <groupId>com.vaadin.demo.apptheme</groupId>
+    <artifactId>extend-existing-app-theme-demo</artifactId>
+    <version>1.0-SNAPSHOT</version>
+    <packaging>jar</packaging>
+
+    <properties>
+        <java.version>11</java.version>
```
Comment (mshabarov) [inline extend-existing-app-theme-demo/pom.xml]: Another use of java 11. Replace with 8 as for other modules?

File: extend-existing-app-theme-demo/pom.xml [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
+
+        <vaadin.version>19.0-SNAPSHOT</vaadin.version>
```
Comment (mshabarov) [inline extend-existing-app-theme-demo/pom.xml]: Now some modules use a snapshots, and some use alphas? Is it better to keep using alphas everywhere in the project?

File: extend-existing-app-theme-demo/pom.xml [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+        <groupId>org.springframework.boot</groupId>
+        <artifactId>spring-boot-starter-parent</artifactId>
+        <version>2.4.0</version>
+    </parent>
+
+    <pluginRepositories>
+        <pluginRepository>
+            <id>vaadin-prereleases</id>
+            <url>https://maven.vaadin.com/vaadin-prereleases</url>
+            <!--<url>http://tools.vaadin.com/nexus/content/repositories/vaadin-prereleases</url>-->
```
Comment (mshabarov) [inline extend-existing-app-theme-demo/pom.xml]: Another commented code which should be removed

File: extend-existing-app-theme-demo/pom.xml [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+            <url>https://repo1.maven.org/maven2/</url>
+            <snapshots><enabled>false</enabled></snapshots>
+        </pluginRepository>
+    </pluginRepositories>
+
+    <repositories>
+        <repository>
+            <id>vaadin-prereleases</id>
+            <url>https://maven.vaadin.com/vaadin-prereleases</url>
+            <!--<url>http://tools.vaadin.com/nexus/content/repositories/vaadin-prereleases</url>-->
```
Comment (mshabarov) [inline extend-existing-app-theme-demo/pom.xml]: Remove commented code

File: extend-existing-app-theme-demo/pom.xml [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+        </dependencies>
+    </dependencyManagement>
+
+    <dependencies>
+        <dependency>
+            <groupId>com.vaadin.demo.apptheme.reusable</groupId>
+            <artifactId>reusable-theme</artifactId>
+            <version>1.0-SNAPSHOT</version>
+        </dependency>
+        <dependency>
```
Comment (mshabarov) [inline extend-existing-app-theme-demo/pom.xml:80]: If you switch to a platform alpha, then you may want to remove flow-server and flow-client deps.

File: extend-existing-app-theme-demo/pom.xml [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+                            <enableAssertions>true</enableAssertions>
+                        </configuration>
+                    </plugin>
+                </plugins>
+            </build>
+        </profile>
+
+
+    </profiles>
+</project>
```
Comment (mshabarov) [inline extend-existing-app-theme-demo/pom.xml:318]: Add a new line

File: extend-existing-app-theme-demo/src/main/java/com/vaadin/demo/apptheme/backend/entity/Company.java [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+  }
+
+  public void setName(String name) {
+    this.name = name;
+  }
+
+  public List<Contact> getEmployees() {
+    return employees;
+  }
+}
```
Comment (mshabarov) [inline extend-existing-app-theme-demo/src/main/java/com/vaadin/demo/apptheme/backend/entity/Company.java:34]: Add a new line

File: extend-existing-app-theme-demo/src/main/java/com/vaadin/demo/apptheme/security/ConfigureUIServiceInitListener.java [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+	 * @param event
+	 *            before navigation event with event details
+	 */
+	private void beforeEnter(BeforeEnterEvent event) {
+		if (!LoginView.class.equals(event.getNavigationTarget())
+		    && !SecurityUtils.isUserLoggedIn()) {
+			event.rerouteTo(LoginView.class);
+		}
+	}
+}
```
Comment (mshabarov) [inline extend-existing-app-theme-demo/src/main/java/com/vaadin/demo/apptheme/security/ConfigureUIServiceInitListener.java:35]: Add a new line

File: extend-existing-app-theme-demo/frontend/themes/specific-theme/components/vaadin-text-field.css (hunk 4 above)
Comment (mshabarov) [inline extend-existing-app-theme-demo/frontend/themes/specific-theme/components/vaadin-text-field.css]:
```suggestion
}

```

File: extend-existing-app-theme-demo/frontend/themes/specific-theme/styles.css (hunk 5 above)
Comment (mshabarov) [inline extend-existing-app-theme-demo/frontend/themes/specific-theme/styles.css]:
```suggestion
}

```

File: extend-existing-app-theme-demo/frontend/themes/specific-theme/theme.json (hunk 6 above)
Comment (mshabarov) [inline extend-existing-app-theme-demo/frontend/themes/specific-theme/theme.json]:
```suggestion
}

```

File: extend-existing-app-theme-demo/pom.xml (hunk 12 above)
Comment (mshabarov) [inline extend-existing-app-theme-demo/pom.xml]:
```suggestion
</project>

```

File: extend-existing-app-theme-demo/src/main/java/com/vaadin/demo/apptheme/backend/entity/Company.java (hunk 13 above)
Comment (mshabarov) [inline extend-existing-app-theme-demo/src/main/java/com/vaadin/demo/apptheme/backend/entity/Company.java]:
```suggestion
}

```

File: extend-existing-app-theme-demo/src/main/java/com/vaadin/demo/apptheme/security/ConfigureUIServiceInitListener.java (hunk 14 above)
Comment (mshabarov) [inline extend-existing-app-theme-demo/src/main/java/com/vaadin/demo/apptheme/security/ConfigureUIServiceInitListener.java]:
```suggestion
}

```

File: extend-existing-app-theme-demo/src/main/java/com/vaadin/demo/apptheme/security/CustomRequestCache.java [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+class CustomRequestCache extends HttpSessionRequestCache {
+
+    @Override
+    public void saveRequest(HttpServletRequest request, HttpServletResponse response) {
+        if (!SecurityUtils.isFrameworkInternalRequest(request)) {
+            super.saveRequest(request, response);
+        }
+    }
+
+}
```
Comment (mshabarov) [inline extend-existing-app-theme-demo/src/main/java/com/vaadin/demo/apptheme/security/CustomRequestCache.java]:
```suggestion
}

```

File: extend-existing-app-theme-demo/src/main/resources/application-prod.properties [hunk 16]
```diff
@@ -0,0 +1,5 @@
+server.port=5000
+spring.datasource.url=jdbc:mysql://${RDS_HOSTNAME}:${RDS_PORT}/${RDS_DB_NAME}
+spring.datasource.username=${RDS_USERNAME}
+spring.datasource.password=${RDS_PASSWORD}
+spring.jpa.hibernate.ddl-auto=create
```
Comment (mshabarov) [inline extend-existing-app-theme-demo/src/main/resources/application-prod.properties]:
```suggestion
spring.jpa.hibernate.ddl-auto=create

```

File: extend-existing-app-theme-demo/src/main/resources/application.properties [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+# spring.datasource.username = <username>
+# spring.datasource.password = <password>
+
+# and replace the h2 dependency in pom.xml with this
+# <dependency>
+#    <groupId>mysql</groupId>
+#    <artifactId>mysql-connector-java</artifactId>
+#    <scope>runtime</scope>
+# </dependency>
+# Further instructions here: https://dzone.com/articles/how-to-connect-a-mysql-database-to-a-vaadin-applic
```
Comment (mshabarov) [inline extend-existing-app-theme-demo/src/main/resources/application.properties]:
```suggestion
# Further instructions here: https://dzone.com/articles/how-to-connect-a-mysql-database-to-a-vaadin-applic

```

File: extend-existing-app-theme-demo/src/test/java/com/vaadin/demo/apptheme/it/AbstractTest.java [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+
+    protected AbstractTest(String route) {
+        this.route = route;
+    }
+
+    private static String getURL(String route) {
+        return String.format("http://%s:%d/%s", SERVER_HOST, SERVER_PORT, route);
+    }
+
+}
```
Comment (mshabarov) [inline extend-existing-app-theme-demo/src/test/java/com/vaadin/demo/apptheme/it/AbstractTest.java]:
```suggestion
}

```

File: flow-in-app-theme-demo/pom.xml [hunk 19]
```diff
@@ -11,15 +11,17 @@
...
+        <maven.compiler.source>1.8</maven.compiler.source>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
         <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
 
         <vaadin.version>19.0-SNAPSHOT</vaadin.version>
 
         <drivers.dir>${project.basedir}/drivers</drivers.dir>
         <drivers.downloader.phase>pre-integration-test</drivers.downloader.phase>
-        <flow.version>6.0.0.alpha2</flow.version>
+        <flow.version>6.0-SNAPSHOT</flow.version>
```
Comment (mshabarov) [inline flow-in-app-theme-demo/pom.xml]: Why snapshots? Platform version is also good to have alpha instead of snapshots

File: reusing-theme-jar-demo/src/main/java/com/vaadin/demo/apptheme/security/ConfigureUIServiceInitListener.java [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+	 * @param event
+	 *            before navigation event with event details
+	 */
+	private void beforeEnter(BeforeEnterEvent event) {
+		if (!LoginView.class.equals(event.getNavigationTarget())
+		    && !SecurityUtils.isUserLoggedIn()) {
+			event.rerouteTo(LoginView.class);
+		}
+	}
+}
```
Comment (mshabarov) [inline reusing-theme-jar-demo/src/main/java/com/vaadin/demo/apptheme/security/ConfigureUIServiceInitListener.java]:
```suggestion
}

```

File: reusing-theme-jar-demo/src/main/java/com/vaadin/demo/apptheme/security/CustomRequestCache.java [hunk 21]
```diff
@@ -0,0 +1,18 @@
...
+class CustomRequestCache extends HttpSessionRequestCache {
+
+    @Override
+    public void saveRequest(HttpServletRequest request, HttpServletResponse response) {
+        if (!SecurityUtils.isFrameworkInternalRequest(request)) {
+            super.saveRequest(request, response);
+        }
+    }
+
+}
```
Comment (mshabarov) [inline reusing-theme-jar-demo/src/main/java/com/vaadin/demo/apptheme/security/CustomRequestCache.java]:
```suggestion
}

```

File: reusing-theme-jar-demo/src/main/resources/application-prod.properties [hunk 22]
```diff
@@ -0,0 +1,5 @@
+server.port=5000
+spring.datasource.url=jdbc:mysql://${RDS_HOSTNAME}:${RDS_PORT}/${RDS_DB_NAME}
+spring.datasource.username=${RDS_USERNAME}
+spring.datasource.password=${RDS_PASSWORD}
+spring.jpa.hibernate.ddl-auto=create
```
Comment (mshabarov) [inline reusing-theme-jar-demo/src/main/resources/application-prod.properties]:
```suggestion
spring.jpa.hibernate.ddl-auto=create

```

File: reusing-theme-jar-demo/src/main/resources/application.properties [hunk 23]
```diff
@@ -0,0 +1,19 @@
...
+# spring.datasource.username = <username>
+# spring.datasource.password = <password>
+
+# and replace the h2 dependency in pom.xml with this
+# <dependency>
+#    <groupId>mysql</groupId>
+#    <artifactId>mysql-connector-java</artifactId>
+#    <scope>runtime</scope>
+# </dependency>
+# Further instructions here: https://dzone.com/articles/how-to-connect-a-mysql-database-to-a-vaadin-applic
```
Comment (mshabarov) [inline reusing-theme-jar-demo/src/main/resources/application.properties]:
```suggestion
# Further instructions here: https://dzone.com/articles/how-to-connect-a-mysql-database-to-a-vaadin-applic

```

File: reusing-theme-jar-demo/src/test/java/com/vaadin/demo/apptheme/it/AbstractTest.java [hunk 24]
```diff
@@ -0,0 +1,51 @@
...
+
+    protected AbstractTest(String route) {
+        this.route = route;
+    }
+
+    private static String getURL(String route) {
+        return String.format("http://%s:%d/%s", SERVER_HOST, SERVER_PORT, route);
+    }
+
+}
```
Comment (mshabarov) [inline reusing-theme-jar-demo/src/test/java/com/vaadin/demo/apptheme/it/AbstractTest.java]:
```suggestion
}

```

**Review by mshabarov — CHANGES_REQUESTED** [review]
Added some other comments related to new lines, snapshot versions and commented code.

After your explanations, I believe it's indeed better to keep the code generally as it is and not to limit it anyhow to save the integrity and similarity of the modules.

One thing I noticed is that the root `pom.xml` is missing. Basically it's not that needed, but may help when you want to build the whole project in one go.

File: extend-existing-app-theme-demo/pom.xml (hunk 7 above)
Comment (mshabarov) [reply extend-existing-app-theme-demo/pom.xml]: `extend-existing-app-theme-demo` does not start because of that.

File: extend-existing-app-theme-demo/pom.xml (hunk 11 above)
Comment (taefi) [reply extend-existing-app-theme-demo/pom.xml:80]: Done.

**Review by mshabarov — APPROVED** [review]
